File: frontend/src/api/knowledge.ts
import apiClient from "./client"

export interface KnowledgeDocument {
  id: string
  title: string
  doc_type: string
  filename: string
  source_url: string
  tags: string[]
  summary: string
  metadata: Record<string, unknown>
  chunk_count: number
  created_at: string
}

export interface KnowledgeChunk {
  id: string
  document_id: string
  chunk_index: number
  content: string
  tags: string[]
  created_at: string
}

export interface SearchResult {
  chunk_id: string
  document_id: string
  content: string
  score: number
  title: string
  doc_type: string
}

export interface KnowledgeStats {
  total_documents: number
  total_chunks: number
  by_type: Record<string, number>
}

export const uploadDocument = (file: File) => {
  const form = new FormData()
  form.append("file", file)
  return apiClient.post("/knowledge/upload", form, {
    headers: { "Content-Type": "multipart/form-data" },
  }).then((r) => r.data)
}

export const indexWebContent = (content: string, filename = "", sourceUrl = "") =>
  apiClient.post("/knowledge/web", { content, filename, source_url: sourceUrl }).then((r) => r.data)

export const searchKnowledge = (query: string, limit = 20, docType?: string) =>
  apiClient.post("/knowledge/search", { query, limit, doc_type: docType }).then((r) => r.data)

export const listDocuments = (docType?: string, limit = 50) =>
  apiClient.get("/knowledge/documents", { params: { doc_type: docType, limit } }).then((r) => r.data)

export const getDocument = (docId: string) =>
  apiClient.get("/knowledge/documents/" + docId).then((r) => r.data)

export const deleteDocument = (docId: string) =>
  apiClient.delete("/knowledge/documents/" + docId).then((r) => r.data)

export const getKnowledgeStats = () =>
  apiClient.get("/knowledge/stats").then((r) => r.data)

export const getKnowledgeContext = (q: string, limit = 5) =>
  apiClient.get("/knowledge/context", { params: { q, limit } }).then((r) => r.data)

export const DOC_TYPES = ["pdf", "markdown", "txt", "code", "web"] as const
export const DOC_TYPE_LABELS: Record<string, string> = {
  pdf: "PDF", markdown: "Markdown", txt: "Text", code: "Code", web: "Web",
}

File: backend/app/api/v1/knowledge.py
"""
Knowledge API - Phase 5.4
Document upload, search, listing for personal knowledge base.
"""

from fastapi import APIRouter, UploadFile, File, Form, Query
from pydantic import BaseModel

from app.knowledge.indexer import KnowledgeIndexer
from app.knowledge.parser import KnowledgeParser, DocumentType
from app.knowledge.retriever import KnowledgeRetriever
from app.utils.logger import get_logger

logger = get_logger(__name__)

router = APIRouter(prefix="/knowledge", tags=["knowledge"])

_indexer: KnowledgeIndexer | None = None
_retriever: KnowledgeRetriever | None = None


def _get_indexer() -> KnowledgeIndexer:
    global _indexer
    if _indexer is None:
        _indexer = KnowledgeIndexer()
    return _indexer


def _get_retriever() -> KnowledgeRetriever:
    global _retriever
    if _retriever is None:
        _retriever = KnowledgeRetriever(_get_indexer())
    return _retriever


class WebContentRequest(BaseModel):
    content: str
    source_url: str = ""
    filename: str = ""


class SearchRequest(BaseModel):
    query: str
    limit: int = 20
    doc_type: str | None = None


@router.post("/upload")
async def upload_document(file: UploadFile = File(...)):
    """Upload and index a document file."""
    try:
        content = (await file.read()).decode("utf-8", errors="replace")
        filename = file.filename or "uploaded"
        indexer = _get_indexer()
        doc = indexer.index(content, filename=filename)
        return {"success": True, "document": doc.to_dict()}
    except Exception as e:
        logger.error("Upload failed", error=str(e))
        return {"success": False, "error": str(e)}


@router.post("/web")
async def index_web_content(req: WebContentRequest):
    """Index web content or text directly."""
    try:
        indexer = _get_indexer()
        doc = indexer.index(req.content, filename=req.filename, source_url=req.source_url)
        return {"success": True, "document": doc.to_dict()}
    except Exception as e:
        logger.error("Web index failed", error=str(e))
        return {"success": False, "error": str(e)}


@router.post("/search")
async def search_knowledge(req: SearchRequest):
    """Search indexed knowledge chunks."""
    retriever = _get_retriever()
    results = retriever.search(req.query, limit=req.limit, doc_type=req.doc_type)
    return {"success": True, "query": req.query, "total": len(results), "results": results}


@router.get("/documents")
async def list_documents(
    doc_type: str | None = Query(default=None),
    limit: int = Query(default=50),
):
    """List indexed documents."""
    indexer = _get_indexer()
    docs = indexer.list_documents(limit=limit, doc_type=doc_type)
    return {"success": True, "total": len(docs), "documents": docs}


@router.get("/documents/{doc_id}")
async def get_document(doc_id: str):
    """Get a document by ID with its chunks."""
    indexer = _get_indexer()
    doc = indexer.get_document(doc_id)
    if not doc:
        return {"success": False, "error": "Document not found"}
    chunks = indexer.get_chunks(doc_id)
    return {
        "success": True,
        "document": doc.to_dict(),
        "chunks": [c.to_dict() for c in chunks],
    }


@router.delete("/documents/{doc_id}")
async def delete_document(doc_id: str):
    """Delete a document and its chunks."""
    indexer = _get_indexer()
    deleted = indexer.delete_document(doc_id)
    if not deleted:
        return {"success": False, "error": "Document not found"}
    return {"success": True, "message": "Document deleted"}


@router.get("/stats")
async def get_stats():
    """Get knowledge base statistics."""
    indexer = _get_indexer()
    return {"success": True, "data": indexer.get_stats()}


@router.get("/context")
async def get_context(q: str = Query(..., description="Task description"), limit: int = 5):
    """Get relevant knowledge context for a task (Agent integration)."""
    retriever = _get_retriever()
    context_text = retriever.get_context_for_task(q, limit=limit)
    return {"success": True, "query": q, "context": context_text}
